### PR TITLE
Clean up output parsing - use common function and remove eval

### PR DESCRIPTION
**Pull request by dennisjlee**
@newville let me know how you feel about this, I'm very open to feedback!

As I started using the library I noticed that there was a lot of repetitive parsing code in this file, and it all relied on `eval` which is less secure and performant. I tried to rewrite in a different way to emphasize safety and performance.

See microbenchmark here:

image

### CODE REVIEW

File: newportxps/XPS_C8_drivers.py [hunk 1]
```diff
@@ -1987,7 +1401,10 @@ def HardwareInternalListGet (self, socketId):
 
     # HardwareDriverAndStageGet :  Smart hardware
     def HardwareDriverAndStageGet (self, socketId, PlugNumber):
-        return self.Send(socketId, 'HardwareDriverAndStageGet(%s, char *, char *)' % str(PlugNumber))
```
Comment (dennisjlee) [inline newportxps/XPS_C8_drivers.py:1990]: This would have parsed the outputs incorrectly before - would have returned one string instead of two.

File: newportxps/XPS_C8_drivers.py [hunk 2]
```diff
@@ -1849,18 +1295,10 @@ def XYZSplineVerification (self, socketId, GroupName, TrajectoryFileName):
...
     def XYZSplineVerificationResultGet (self, socketId, PositionerName):
-        command = 'XYZSplineVerificationResultGet(' + PositionerName + ',char *,double *,double *,double *,double *)'
+        outputs = XPSOutputs('char', 'double', 'double', 'double', 'double')
+        command = f'XYZSplineVerificationResultGet({PositionerName},{outputs})'
         error, returnedString = self.Send(socketId, command)
-        if (error != 0):
-            return [error, returnedString]
-
-        i, j, retList = 0, 0, [error]
-        for paramNb in range(4):
```
Comment (dennisjlee) [inline newportxps/XPS_C8_drivers.py:1858]: this was previously parsing the wrong number of outputs

File: newportxps/XPS_C8_drivers.py [hunk 3]
```diff
@@ -1773,39 +1252,22 @@ def XYLineArcVerification (self, socketId, GroupName, TrajectoryFileName):
...
     def XYLineArcParametersGet (self, socketId, GroupName):
-        command = 'XYLineArcParametersGet(' + GroupName + ',char *,double *,double *,int *)'
+        outputs = XPSOutputs('char', 'double', 'double', 'int')
+        command = f'XYLineArcParametersGet({GroupName},{outputs})'
         error, returnedString = self.Send(socketId, command)
-        if (error != 0):
-            return [error, returnedString]
-
-        i, j, retList = 0, 0, [error]
-        for paramNb in range(3):
```
Comment (dennisjlee) [inline newportxps/XPS_C8_drivers.py:1803]: this was parsing the wrong number of outputs

File: newportxps/XPS_C8_drivers.py [hunk 4]
```diff
@@ -1773,39 +1252,22 @@ def XYLineArcVerification (self, socketId, GroupName, TrajectoryFileName):
...
     def XYLineArcVerificationResultGet (self, socketId, PositionerName):
-        command = 'XYLineArcVerificationResultGet(' + PositionerName + ',char *,double *,double *,double *,double *)'
+        outputs = XPSOutputs('char', 'double', 'double', 'double', 'double')
+        command = f'XYLineArcVerificationResultGet({PositionerName},{outputs})'
         error, returnedString = self.Send(socketId, command)
-        if (error != 0):
-            return [error, returnedString]
-
-        i, j, retList = 0, 0, [error]
-        for paramNb in range(4):
```
Comment (dennisjlee) [inline newportxps/XPS_C8_drivers.py:1782]: this was parsing the wrong number of outputs

File: newportxps/XPS_C8_drivers.py [hunk 5]
```diff
@@ -1579,31 +1123,17 @@ def PositionerDACOffsetDualSet (self, socketId, PositionerName, PrimaryDACOffset
 
     # PositionerCorrectorAutoTuning :  Astrom&Hagglund based auto-tuning
     def PositionerCorrectorAutoTuning (self, socketId, PositionerName, TuningMode):
-        command = 'PositionerCorrectorAutoTuning(' + PositionerName + ',' + str(TuningMode) + ',double *,double *,double *)'
-        return self.Send(socketId, command)
```
Comment (dennisjlee) [inline newportxps/XPS_C8_drivers.py:1583]: this had been returning early before and not parsing the outputs at all

**Review by dennisjlee — COMMENTED** [review]
I've noted below a few places below where it looked like the outputs were mis-parsed to me previously.

File: newportxps/XPS_C8_drivers.py [hunk 6]
```diff
@@ -58,20 +89,20 @@ def wrapper(*args, **kw):
...
         except socket.error as err: #  (errNb, errString):
             print( 'Socket error : ', err.errno, err)
-            return [-2, '']
+            return -2, ''
 
-        for i in range(len(ret)):
-            if (ret[i] == ','):
-                return [int(ret[0:i]), ret[i+1:-9]]
+        error, rest = ret[:-len(suffix)].split(',', 1)
+        return int(error), rest
```
Comment (dennisjlee) [inline newportxps/XPS_C8_drivers.py:105]: sorry there was one unrelated cleanup in this function - I updated it to return a tuple instead of a list (easier for the type checker to understand when the first and second item have different types), and I also eliminated this loop so that there was no possible path of returning `None` (also for the type checker)

**Comment by newville** [comment]
@dennisjlee Yeah, this file is derived from the original code provided by Newport.   It was always a bit ugly and had to be altered to support both Py2 and Py3 for a while.  It has not aged well.   I think it would benefit from cleanup and refactoring.  Like, using f-strings would be very helpful.   We don't have a good automated test suite or process, but I think that should not necessarily prevent refactoring.

**Comment by dennisjlee** [comment]
Cool thanks for the context @newville. I used f-strings in places that I touched and also made an effort at adding type annotations, but I could definitely do more in the future. I saw in the pyproject.toml that this project supports Python 3.8+ so I used that when developing and made sure to use type annotations that are compatible with 3.8.

**Comment by newville** [comment]
@dennisjlee I am in a better place to look at this, so I am going to merge this (and then test it locally a bunch).  I'm hoping to tag a release of 1.0.0 soon-ish.

**Comment by dennisjlee** [comment]
@newville thank you very much! Let me know if you have any feedback or find any bugs, happy to follow up.

I haven't had much time to look more at this codebase lately but one thing I've been pondering in the back of my head is how to make these functions in XPS_C8_drivers.py have correct return types annotated (automatically if possible). If I have more time I'll see if I can tinker with that further.
